File: doc/codes/python/examples/api/command/pattern-example.py
import asyncio

import numpy as np

from autd3 import Client, ClientConfig
from autd3.commands import Pattern
from autd3.geometry import Autd3, Geometry
from autd3.units import m, s
from autd3_link_nop import Nop
from autd3_pattern import FocusOption, focus, wavelength


async def main() -> None:
    geometry = Geometry([Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0])])
    client = await Client.open(geometry, Nop(), ClientConfig())

    emissions = geometry.pattern_buffer()
    focus(
        geometry,
        geometry.center() + np.array([0.0, 0.0, 150.0]),
        wavelength(340 * m / s),
        FocusOption(),
        emissions,
    )

    builder = client.datagram_builder()
    builder.push(Pattern(emissions))
    frames = builder.build()
    for frame in frames:
        await client.send_checked(frame)

    await client.close()


asyncio.run(main())

File: doc/codes/rust/examples/api/command/pattern-example.rs
use anyhow::Result;

use autd3_rs::commands::Pattern;
use autd3_rs::geometry::{Autd3, Geometry, offset};
use autd3_rs::units::{m, mm, s};
use autd3_rs::{Client, ClientConfig};
use autd3_rs_link_nop::Nop;
use autd3_rs_pattern::{FocusOption, focus, wavelength};

#[tokio::main(flavor = "multi_thread")]
async fn main() -> Result<()> {
    let geometry = Geometry::new(vec![Autd3::default()]);
    let client = Client::open(&geometry, Nop, ClientConfig::default()).await?;

    let mut emissions = geometry.pattern_buffer();
    focus(
        &geometry,
        geometry.center() + offset(0.0 * mm, 0.0 * mm, 150.0 * mm),
        wavelength(340.0 * m / s),
        &FocusOption::default(),
        &mut emissions,
    );

    let mut builder = client.datagram_builder();
    builder.push(Pattern::new(&emissions));
    let frames = builder.build()?;
    for frame in &frames {
        client.send_checked(frame).await?;
    }

    client.close().await?;
    Ok(())
}
